stdlib/public/SDK/Foundation/JSONEncoder.swift

```diff
@@ -0,0 +1,1743 @@
+    fileprivate struct _Strategies {
+        let dateDecodingStrategy: DateDecodingStrategy
+        let dataDecodingStrategy: DataDecodingStrategy
+        let nonConformingFloatDecodingStrategy: NonConformingFloatDecodingStrategy
+    }
+
+    /// The strategies set on the top-level decoder.
+    fileprivate var strategies: _Strategies {
+        return _Strategies(dateDecodingStrategy: dateDecodingStrategy,
+                           dataDecodingStrategy: dataDecodingStrategy,
+                           nonConformingFloatDecodingStrategy: nonConformingFloatDecodingStrategy)
+    }
+
+    // MARK: - Constructing a JSON Encoder
+
+    /// Initializes `self` with default strategies.
+    public init() {}
+
+    // MARK: - Decoding Values
+
+    /// Decodes a top-level value of the given type from the given JSON representation.
+    ///
+    /// - parameter type: The type of the value to decode.
+    /// - parameter data: The data to decode from.
+    /// - returns: A value of the requested type.
+    /// - throws: `CocoaError.coderReadCorrupt` if values requested from the payload are corrupted, or if the given data is not valid JSON.
+    /// - throws: An error if any value throws an error during decoding.
+    open func decode<Value : Codable>(_ type: Value.Type, from data: Data) throws -> Value {
+        let topLevel = try JSONSerialization.jsonObject(with: data)
+        let decoder = _JSONDecoder(storage: topLevel, strategies: self.strategies)
+        return try Value(from: decoder)
+    }
+}
+
+// MARK: -
+
+fileprivate class _JSONDecoder : Decoder, SingleValueDecodingContainer {
+    /// The storage to decode values from.
+    let storage: Any
+
+    /// The path of coding keys taken to get to this point in encoding.
+    let codingKeyContext: [CodingKey]
+
+    /// Strategies to use in converting values for encoding.
+    let strategies: JSONDecoder._Strategies
+
+    /// Initializes `self` with the given storage and strategies.
+    ///
+    /// - parameter storage: The value to reference for decoding.
+    /// - parameter strategies: The strategies to use while decoding.
+    /// - parameter context: The decoding context we're working in.
+    init(storage: Any, strategies: JSONDecoder._Strategies, context: [CodingKey] = []) {
+        self.storage = storage
+        self.strategies = strategies
+        self.codingKeyContext = context
+    }
+
+    // MARK: - Decoder Methods
+
+    func container<Key : CodingKey>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> {
+        guard storage is [Any] || storage is [String : Any] else {
+            throw CocoaError.expectedContainer(in: codingKeyContext, reality: storage)
+        }
+
+        return _JSONKeyedDecodingContainer(referencing: self)
+    }
+
+    func singleValueContainer() throws -> SingleValueDecodingContainer {
+        return self
+    }
+
+    // MARK: - SingleValueDecodingContainer Methods
+
+    func decode(_ type: Bool.Type) throws -> Bool {
```

Comment:
You can overload on return value in Swift - no need to take the metatype parameter